### PR TITLE
Feature/permutation

### LINKED ISSUE #1526: [feature request] random permutation
It would be nice if we could get `jax.random` functions analogous to `np.random.permutation` and `np.random.shuffle`.

### PR DESCRIPTION
**Pull request by MichaelMarien**
fixes #1526

### CODE REVIEW

File: jax/random.py [hunk 1]
```diff
@@ -367,6 +367,24 @@ def shuffle(key, x, axis=0):
...
+  Args:n
+    key: a PRNGKey used as the random key.
+    x: the array or integer range to be shuffled.
+
+  Returns:
+    A shuffled version of x or array range
+  """
+  if isinstance(x, (int, onp.integer)):
+    arr = onp.arange(x)
+  else:
```
Comment (jekbradbury) [inline jax/random.py]: Don't you want this shuffle to occur regardless of whether the array comes from an `arange` call?

Comment (MichaelMarien) [reply jax/random.py]: Absolutely, thanks for that. I forgot the fix the test, danger of copy-pasting.

Comment (mattjj) [reply jax/random.py]: In the spirit of making this API slightly more explicit and less magically-polymorphic than NumPy's API, what if we just required the user to pass an array for `x`, i.e. require them to write `np.arange` when needed?

Comment (mattjj) [reply jax/random.py]: Nevermind, I guess that's the whole point of this wrapper!

**Comment by grisaitis** [comment]
anything holding this back? just came looking for this feature. 

also, design q - instead of `permutation` wrapping `shuffle`, what about the other way around? the sort key logic in `shuffle` seems (to me) like permuting an `arange`.

**Comment by mattjj** [comment]
Thanks for pinging this. We've been swamped and as a result have dropped the ball on a lot of good contributions :/

I think the `isinstance(x, (int, onp.integer))` check won't work with Tracer instances. We should probably instead switch on the shape, i.e. whether `onp.shape(x) == ()`.

I haven't reviewed the testing code yet.

@jakevdp since you've been thinking about this stuff a little recently, are you interested in looking at this and making sure it's consistent with your plans? (Though don't take this on if you're too busy!)

**Comment by mattjj** [comment]
I might be mistaken about the `isinstance` because it seems to be working in the tests (?).

The tests LGTM, modulo I'd like to understand the `isinstance` behavior better.

**Comment by mattjj** [comment]
Oh, duh. If it's an integer, we need it to be static.

I have some ideas for tweaks, let me see if I can put a commit on top of this.

**Comment by jakevdp** [comment]
I think this may suffer from the same bug reported in #2066; for multi-dimensional arrays, the ``_shuffle`` function shuffles each subarray independently rather than as a group as in ``np.permutation``. For example:
```python
>>> import numpy as np
>>> x = np.arange(25).reshape(5, 5)
>>> np.random.permutation(x)
array([[15, 16, 17, 18, 19],
       [ 5,  6,  7,  8,  9],
       [ 0,  1,  2,  3,  4],
       [20, 21, 22, 23, 24],
       [10, 11, 12, 13, 14]])

>>> import jax.random
>>> jax.random._shuffle(jax.random.PRNGKey(0), x, 0)
DeviceArray([[10, 21, 22,  8, 14],
             [ 5, 11, 12,  3, 24],
             [20,  1, 17, 18,  4],
             [ 0,  6,  7, 23,  9],
             [15, 16,  2, 13, 19]], dtype=int32)
```
I think the solution is probably to do something similar to that in #2790 for multi-dimensional inputs.

**Comment by mattjj** [comment]
@jakevdp once #2790 lands, can we just update the function in this PR to call `shuffle` with `axis=None` and get things to work right?

**Comment by mattjj** [comment]
(Wow, @hawkinsp told me that I can push to PR branches, and he was right! woo!)

**Comment by mattjj** [comment]
I rebased the commits here onto master, and also tweaked some of the logic to raise better errors.

**Comment by mattjj** [comment]
I think I'll just make this function raise an error on >1d inputs for now.

**Comment by mattjj** [comment]
Thanks so much for the contribution, @MichaelMarien ! Sorry for the glacially slow response on our end, but we got to the right place eventually. (Thanks also @grisaitis for the ping.)

@jakevdp could you review this PR (after my changes just now) and LMK what you think? In particular, by raising an error on the >1d case, are we set up to integrate with #2790 well?

File: tests/random_test.py [hunk 2]
```diff
@@ -169,6 +170,46 @@ def testShuffle(self, dtype):
...
+    key = random.PRNGKey(0)
+    x = onp.arange(100).astype(dtype)
+    rand = lambda key: random.permutation(key, x)
+    crand = api.jit(rand)
+
+    perm1 = rand(key)
+    perm2 = crand(key)
+
+    self.assertTrue(onp.all(perm1 == perm2))
+    self.assertTrue(onp.all(perm1.dtype == perm2.dtype))
```
Comment (jakevdp) [inline tests/random_test.py:186]: Drop `onp.all` here.

Also, these two asserts could be replaced by one `assertArraysAllClose` with `check_dtypes=True`.

File: tests/random_test.py [hunk 3]
```diff
@@ -169,6 +170,46 @@ def testShuffle(self, dtype):
...
+    key = random.PRNGKey(0)
+    x = 100
+    rand = lambda key: random.permutation(key, x)
+    crand = api.jit(rand)
+
+    perm1 = rand(key)
+    perm2 = crand(key)
+
+    self.assertTrue(onp.all(perm1 == perm2))
+    self.assertTrue(onp.all(perm1.dtype == perm2.dtype))
```
Comment (jakevdp) [inline tests/random_test.py:202]: Same here: `assertArraysAllClose`

**Comment by jakevdp** [comment]
I think raising an error for >1D is fine for now. I can wrap the rest into #2790, since it's quite similar.

File: tests/random_test.py (hunk 2 above)
Comment (mattjj) [reply tests/random_test.py:186]: Good catches! Thanks.

Comment (mattjj) [reply tests/random_test.py:186]: Done!

File: tests/random_test.py (hunk 3 above)
Comment (mattjj) [reply tests/random_test.py:202]: Done!

**Comment by mattjj** [comment]
Internal tests pass!
